File: Sample-ObjC/06-query-data/Mesosfer-ObjC/QueryDataViewController.m
/**
 * Copyright (c) 2016, Mesosfer
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "QueryDataViewController.h"
#import "UIViewController+Helper.h"
#import <Mesosfer/Mesosfer.h>

@interface QueryDataViewController ()

@property (nullable, nonatomic, strong) NSMutableArray<MFData *>* datas;

@end

@implementation QueryDataViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.tableView.estimatedRowHeight = 100;
    self.tableView.rowHeight = UITableViewAutomaticDimension;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self queryData];
}

- (void)queryData {
    NSString *bucket = @"Beacon";
    MFQuery *query = [MFData queryWithBucket:bucket];
    [query findAsyncWithBlock:^(NSArray * _Nullable objects, NSError * _Nullable error) {
        if (error) {
            [self showErrorWithTitle:@"Query data failed" error:error handler:nil];
            return;
        }
        
        if (objects) {
            self.title = [NSString stringWithFormat:@"Query '%@'", bucket];
            self.datas = [NSMutableArray arrayWithArray:objects];
            [self.tableView reloadData];
        }
    }];
}

#pragma mark - UITableViewDelegate

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    if (self.datas) {
        return self.datas.count;
    }
    
    return 0;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    if (self.datas) {
        return self.datas[section].objectId;
    }
    return nil;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cellData" forIndexPath:indexPath];
    
    if (self.datas) {
        MFData *data = self.datas[indexPath.section];
        
        if ([data dictionary]) {
            cell.textLabel.text = [NSString stringWithFormat:@"%@", [data dictionary]];
        }
    }
    
    return cell;
}

@end

File: Sample-Swift/06-query-data/Mesosfer-Swift/QueryDataViewController.swift
/**
 * Copyright (c) 2016, Mesosfer
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation
import UIKit
import Mesosfer

class QueryDataViewController: UITableViewController {
    var datas: [MFData]?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.estimatedRowHeight = 100
        self.tableView.rowHeight = UITableViewAutomaticDimension
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        self.querydata()
    }
    
    func querydata() {
        let bucket = "Beacon"
        MFData.query(withBucket: bucket).findAsync { (datas, error) in
            if let e = error as? NSError {
                self.showError(title: "Query data failed", error: e)
                return
            }
            
            if let datas = datas as? [MFData] {
                self.title = "Query '\(bucket)'"
                self.datas = datas
                self.tableView.reloadData()
            }
        }
    }
}

extension QueryDataViewController {
    override func numberOfSections(in tableView: UITableView) -> Int {
        if let datas = self.datas {
            return datas.count
        }
        return 0
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 1
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        if let datas = self.datas {
            return datas[section].objectId
        }
        return nil
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "cellData", for: indexPath)
        
        if let datas = self.datas {
            let data = datas[indexPath.section]
            
            if let dict = data.dictionary() as? [String:AnyObject] {
                cell.textLabel?.text = "\(dict)"
            }
        }
        
        return cell
    }
}
